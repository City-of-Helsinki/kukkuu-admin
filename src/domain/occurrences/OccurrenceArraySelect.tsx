import React, { ChangeEvent } from 'react';
import {
  SelectArrayInput,
  ReferenceArrayInput,
  useTranslate,
  useInput,
} from 'react-admin';
import { useFormContext } from 'react-hook-form';

import { Occurrence_occurrence as Occurrence } from '../../api/generatedTypes/Occurrence';
import { toShortDateTimeString } from '../../common/utils';

const AllChoice = ({
  children,
  choices,
  allChoiceLabel,
  getChoices,
  ...rest
}: any) => {
  return React.cloneElement(children, {
    ...rest,
    choices: [{ id: 'all', name: allChoiceLabel }, ...getChoices(choices)],
  });
};

export function getFilters(
  eventId?: string
): Record<string, string> | undefined {
  if (!eventId) {
    return;
  }

  return {
    eventId,
  };
}

export const getChoices = (records: Occurrence[]) => {
  return records.map(({ id, time }) => ({
    id,
    name: toShortDateTimeString(new Date(time)),
  }));
};

export const getValues = (previousValues: string[], nextValues: string[]) => {
  const wasAllSelected = previousValues.includes('all');
  const willHaveOtherValueThanAll = nextValues.length > 0;
  const willHaveNoValue = nextValues.length === 0;
  const allWillBeAdded =
    !previousValues.includes('all') && nextValues.includes('all');

  if (wasAllSelected && willHaveOtherValueThanAll) {
    return nextValues.filter((value: string) => value !== 'all');
  } else if (willHaveNoValue || allWillBeAdded) {
    return ['all'];
  } else {
    return nextValues;
  }
};

type ReferenceArrayInputProps = Parameters<typeof ReferenceArrayInput>[0];
type Props = Omit<ReferenceArrayInputProps, 'children' | 'reference'> & {
  source: string;
  eventId?: string;
  allText: string;
};

const OccurrenceArraySelect = ({ eventId, allText, ...rest }: Props) => {
  const filter = getFilters(eventId);
  const {
    field: { value, name },
  } = useInput({ source: rest.source });
  const form = useFormContext();
  const t = useTranslate();

  // By default the multi select component does not explicitly have an
  // "all" option. This custom handler implements behavior for the all
  // option.
  // - when the user deselects all choices, all is selected
  // - when the user selects some (other) option (than all), "all"
  //   selection is removed
  // - when the user selects "all", all other choices are removed
  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const previousValues = value;
    // The value is actually an array because we are using a
    // multi select
    const nextValues = e.target.value as unknown as string[];
    const values = getValues(previousValues, nextValues);

    form.setValue(name, values);
  };

  return (
    <ReferenceArrayInput
      {...rest}
      resource="occurrences"
      reference="occurrences"
      filter={filter}
      onChange={handleChange}
    >
      <AllChoice allChoiceLabel={t(allText)} getChoices={getChoices}>
        <SelectArrayInput />
      </AllChoice>
    </ReferenceArrayInput>
  );
};

export default OccurrenceArraySelect;
