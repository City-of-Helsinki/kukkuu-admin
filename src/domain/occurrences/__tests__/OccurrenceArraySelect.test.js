import React from 'react';
import { render } from '@testing-library/react';
import { TestContext, SimpleForm } from 'react-admin';

import OccurrenceArraySelect, {
  getChoices,
  getFilters,
  getValues,
} from '../OccurrenceArraySelect';

describe('<OccurrenceArraySelect />', () => {
  it('should render without errors', () => {
    jest.spyOn(global.console, 'error').mockImplementationOnce(() => null);

    expect(
      render(
        <TestContext
          initialState={{ admin: { resources: { occurrences: { data: {} } } } }}
        >
          <SimpleForm>
            <OccurrenceArraySelect
              allText="allText"
              eventId="1"
              source="event"
            />
          </SimpleForm>
        </TestContext>
      )
    ).toMatchSnapshot();
  });
});

describe('OccurrenceArraySelect helpers', () => {
  describe('getFilters', () => {
    it('should return undefined if there is no event id', () => {
      expect(getFilters()).toBeUndefined();
    });

    it('otherwise it should return filters', () => {
      expect(getFilters('123')).toEqual({
        eventId: '123',
      });
    });
  });

  describe('getChoices', () => {
    it('should make choices out of occurrences', () => {
      const occurrences = [
        {
          id: '123',
          time: '2020-10-7',
        },
        {
          id: '12',
          time: '2020-11-7',
        },
      ];
      const choices = getChoices(occurrences);

      expect(choices.length).toEqual(occurrences.length);
      expect(choices[0]).toMatchInlineSnapshot(`
        Object {
          "id": "123",
          "name": "7.10.2020, 00.00",
        }
      `);
    });
  });

  describe('getValues', () => {
    it('should remove all from selected values if some other option is selected', () => {
      expect(getValues(['all'], ['other options']).includes('all')).toBe(false);
    });

    it('should only return all if all is selected', () => {
      expect(getValues(['value 1', 'value 1'], ['all'])).toEqual(['all']);
    });

    it('should return all if nothing is selected', () => {
      expect(getValues(['value 1', 'value 1'], [])).toEqual(['all']);
    });

    it('otherwise it should return next values', () => {
      const nextValues = ['value 1', 'value 2'];

      expect(getValues(['value 1'], nextValues)).toEqual(nextValues);
    });
  });
});
