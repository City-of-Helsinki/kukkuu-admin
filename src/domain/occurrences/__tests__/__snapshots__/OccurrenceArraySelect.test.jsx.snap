// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OccurrenceArraySelect /> > should render without errors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="MuiCardContent-root css-1fd8x6x-MuiCardContent-root-RaSimpleForm-root"
        >
          <div
            class="MuiStack-root css-1a69w1n-MuiStack-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginDense ra-input ra-input-event css-ljq1zz-MuiFormControl-root-RaSelectArrayInput-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="event"
                id="event-outlined-label"
              >
                <span>
                  resources.undefined.fields.event
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1elr7vf-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
                data-testid="selectArray"
              >
                <div
                  aria-controls=":r0:"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="event-outlined-label event"
                  class="MuiSelect-select MuiSelect-filled MuiSelect-multiple MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1hejy8i-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
                  id="event"
                  role="combobox"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  name="event"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
              >
                <span>
                  ​
                </span>
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular RaToolbar-desktopToolbar css-qmvuda-MuiToolbar-root-RaToolbar-root"
          role="toolbar"
        >
          <div
            class="RaToolbar-defaultToolbar"
          >
            <button
              aria-label="ra.action.save"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-t039ty-MuiButtonBase-root-MuiButton-root-RaSaveButton-root"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SaveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3m3-10H5V5h10z"
                />
              </svg>
              ra.action.save
            </button>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="MuiCardContent-root css-1fd8x6x-MuiCardContent-root-RaSimpleForm-root"
      >
        <div
          class="MuiStack-root css-1a69w1n-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-marginDense ra-input ra-input-event css-ljq1zz-MuiFormControl-root-RaSelectArrayInput-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="event"
              id="event-outlined-label"
            >
              <span>
                resources.undefined.fields.event
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1elr7vf-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
              data-testid="selectArray"
            >
              <div
                aria-controls=":r0:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="event-outlined-label event"
                class="MuiSelect-select MuiSelect-filled MuiSelect-multiple MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1hejy8i-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
                id="event"
                role="combobox"
                tabindex="0"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                name="event"
                tabindex="-1"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
            >
              <span>
                ​
              </span>
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular RaToolbar-desktopToolbar css-qmvuda-MuiToolbar-root-RaToolbar-root"
        role="toolbar"
      >
        <div
          class="RaToolbar-defaultToolbar"
        >
          <button
            aria-label="ra.action.save"
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-t039ty-MuiButtonBase-root-MuiButton-root-RaSaveButton-root"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SaveIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3m3-10H5V5h10z"
              />
            </svg>
            ra.action.save
          </button>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
